pimd/pim6_mld.c

```diff
@@ -0,0 +1,2705 @@
+	struct gm_if *gm_ifp;
+
+	if (!sub)
+		return;
+	gm_ifp = sub->iface;
+
+	*subp = NULL;
+	sub->refcount--;
+
+	if (sub->refcount)
+		return;
+
+	gm_subscribers_del(gm_ifp->subscribers, sub);
+	XFREE(MTYPE_GM_SUBSCRIBER, sub);
+}
+
+/****************************************************************************/
+
+/* bundle query timer values for combined v1/v2 handling */
+struct gm_query_timers {
+	unsigned qrv;
+	unsigned max_resp_ms;
+	unsigned qqic_ms;
+
+	struct timeval fuzz;
+	struct timeval expire_wait;
+};
+
+static void gm_expiry_calc(struct gm_query_timers *timers)
+{
+	unsigned expire =
+		(timers->qrv - 1) * timers->qqic_ms + timers->max_resp_ms;
+	ldiv_t exp_div = ldiv(expire, 1000);
+
+	timers->expire_wait.tv_sec = exp_div.quot;
+	timers->expire_wait.tv_usec = exp_div.rem * 1000;
+	timeradd(&timers->expire_wait, &timers->fuzz, &timers->expire_wait);
+}
+
+static void gm_sg_free(struct gm_sg *sg)
+{
+	/* t_sg_expiry is handled before this is reached */
+	THREAD_OFF(sg->t_sg_query);
+	gm_packet_sg_subs_fini(sg->subs_negative);
+	gm_packet_sg_subs_fini(sg->subs_positive);
+	XFREE(MTYPE_GM_SG, sg);
+}
+
+/* clang-format off */
+static const char *const gm_states[] = {
+	[GM_SG_NOINFO]			= "NOINFO",
+	[GM_SG_JOIN]			= "JOIN",
+	[GM_SG_JOIN_EXPIRING]		= "JOIN_EXPIRING",
+	[GM_SG_PRUNE]			= "PRUNE",
+	[GM_SG_NOPRUNE]			= "NOPRUNE",
+	[GM_SG_NOPRUNE_EXPIRING]	= "NOPRUNE_EXPIRING",
+};
+/* clang-format on */
+
+CPP_NOTICE("TODO: S,G entries in EXCLUDE (i.e. prune) unsupported");
+/* tib_sg_gm_prune() below is an "un-join", it doesn't prune S,G when *,G is
+ * joined.  Whether we actually want/need to support this is a separate
+ * question - it is almost never used.  In fact this is exactly what RFC5790
+ * ("lightweight" MLDv2) does:  it removes S,G EXCLUDE support.
+ */
+
+static void gm_sg_update(struct gm_sg *sg, bool has_expired)
+{
+	struct gm_if *gm_ifp = sg->iface;
+	enum gm_sg_state prev, desired;
+	bool new_join;
+	struct gm_sg *grp = NULL;
+
+	if (!pim_addr_is_any(sg->sgaddr.src))
+		grp = gm_sg_find(gm_ifp, sg->sgaddr.grp, PIMADDR_ANY);
+	else
+		assert(sg->state != GM_SG_PRUNE);
+
+	if (gm_packet_sg_subs_count(sg->subs_positive)) {
+		desired = GM_SG_JOIN;
+		assert(!sg->t_sg_expire);
+	} else if ((sg->state == GM_SG_JOIN ||
+		    sg->state == GM_SG_JOIN_EXPIRING) &&
+		   !has_expired)
+		desired = GM_SG_JOIN_EXPIRING;
+	else if (!grp || !gm_packet_sg_subs_count(grp->subs_positive))
+		desired = GM_SG_NOINFO;
+	else if (gm_packet_sg_subs_count(grp->subs_positive) ==
+		 gm_packet_sg_subs_count(sg->subs_negative)) {
+		if ((sg->state == GM_SG_NOPRUNE ||
+		     sg->state == GM_SG_NOPRUNE_EXPIRING) &&
+		    !has_expired)
+			desired = GM_SG_NOPRUNE_EXPIRING;
+		else
+			desired = GM_SG_PRUNE;
+	} else if (gm_packet_sg_subs_count(sg->subs_negative))
+		desired = GM_SG_NOPRUNE;
+	else
+		desired = GM_SG_NOINFO;
+
+	if (desired != sg->state && !gm_ifp->stopping) {
+		zlog_debug(log_sg(sg, "%s => %s"), gm_states[sg->state],
```

Comment:
Self review note: I still need to put debug guards on all the `zlog_debug` statements, I just haven't gotten around to it.

Reply:
(Done)